### ISSUE TITLE
TemplateCache should be async

### ISSUE THREAD

**Issue by megawac** [marionettejs/backbone.marionette#2237]
I'm heading for dinner so I'm not about to hunt down all the relevant issues (feel free to edit them in if you are in the mood for a scavenger hunt)

`getTemplate` should be async by default which would allow users to download items from the server. This is spurred again by @Tvrqvoise comments on the gitter channel

> I use Dustjs, which by default renders asynchronously. Which means that it can load templates on the fly (so you don't have to load every single page on your site just to show one).
> I can make it sync, which works for my needs, but I'd like to have a dust adaptor which can handle either.
> (Also, I'm thinking I could just have the callback from sync call render again?)

---

This also fits well with @thejameskyle's sentiments of async rendering (which makes a log of sense). If `getTemplate` becomes async, `render` will have to be async as well.

Conceptual use:

``` js
import {ItemView} from 'marionette';
import $ from 'jquery'

export default class View extends ItemView {
    get template() {
        // Template cache would compile the template when the deferred resolves
        return $.ajax('./view-template.tpl');
    }
}
```

**Comment by jmeas** [comment]
> a log of sense

:evergreen_tree:

**Comment by thejameskyle** [comment]
:+1: 

![log-of-sense](https://cloud.githubusercontent.com/assets/952783/5985332/8a1325aa-a895-11e4-923e-17d9a7249f8d.jpg)

**Comment by paulfalgout** [comment]
I could have used this early on.  

When we first started converting our reasonably large app from mainly server generated DOM to Marionette,  we had quite a few "templates" that were partially generated from the server.  I had to extend `ItemView` quite a bit to get it to work.  Though we were often having to get multiple templates in single calls from the server and caching them for use in other views later.

However I hope to never have to load a template async when I need to render something ever again.  I much prefer to load potentially needed templates ahead of renders.

**Comment by Tvrqvoise** [comment]
The biggest hurdle in this is not just building TemplateCache to be async, but making ```render``` async (which we'd need to do to get TemplateCache async).  This is challenging for two reasons - firstly because ```render``` is used often synchronously in our own code.  Secondly (and worse) because a lot of components which people tend to overwrite (```RegionManager```, ```attachElContent```, etc) may expect the DOM to be ready as soon as ```render``` is called.

That being said async ```render``` is, I think, worth it.  Not only would this allow for async template loading,  but also for other neat stuff such as buffering DOM updates to requestAnimationFrame, in the way that React does.

**Comment by Tvrqvoise** [comment]
As pointed out by @somebody32 on the gitter channel, it looks like we used to have this feature: https://github.com/marionettejs/backbone.marionette.async

**Comment by jmeas** [comment]
Tbqh I'm not that big a fan of this. Forcing everyone to deal with async render sucks, especially if it's to support a use case that isn't a best practice.

Imo people should not be downloading templates from the server. Rather, they should precompile them and bundle them with their build tool.

I'd rather we just do two things:

1. Add a note in the documentation encouraging people to not do this
2. Maintain some shim that lets you do this, but is not in core

@Tvrqvoise pointed out that marionette.async repo that does this. Maybe we can just put the shim there?

**Comment by Tvrqvoise** [comment]
Yeah, the deprecation message in that marionette.async repo makes me feel concerned about having this feature in core.  Maybe today's code base is different enough that we could support async render without the same overhead, but who knows.

That being said, let's really talk about what we'd be gaining from this before we decide to axe it.  I don't think server loading is the biggest benefit here: async render adds a lot of flexibility to what a renderer can do just in general.  All the other big dogs (React, Angular and I think even Ember and Meteor) render asynchronously, and that allows them to _smoke_ us from a layout performance perspective.  Obviously, that's very much a per-app type thing, but using a system which enforces synchronous rendering inherently prevents you from making certain optimizations.  Adding async render would also allow us to use templating libraries which async render by default (eg. React), or do lazy-loading of templating files.  The best part of this is that _Marionette has a great API for this already in life cycle events._  Users who are making heavy use of life cycle events might not even notice that this change occurred, and it provides a much cleaner declarative alternative to .then'ing everything.

On the other hand, the maintenance burden is potentially very large.  It would be one more thing to worry about when doing the 2.x => 3.x migration, and would be hard to fix via an upgrade script or the like.  Very much one of the benefits of Marionette is that it's easy to reason about, and if this means losing that, I'm not sure if it's worth it.

Ultimately, I'm on the fence.  For what it's worth, I think it would be cool to have async render, and I know that people at company would make heavy use of it.  On the other hand, I haven't contributed enough to Marionette or been around enough to say if this is good thing for the long term health of the library.

+ $0.02

**Comment by jmeas** [comment]
> All the other big dogs (React, Angular and I think even Ember and Meteor) render asynchronously, and that allows them to smoke us from a layout performance perspective

All things equal, when comparing sync render and async render there's no competition: synchronous will be faster. This is because Promises are substantially slower than plain function calls. I have a vague recollection of some issue on this repo where someone adamantly argued against async ItemView rendering because of the huge performance loss in the CollectionView.

It's not too tough to create a bespoke experiment that shows this:

```js
var start = performance.now();

// Use "Promise.resolve" to mimic an async render method
// that resolves instantly. After it renders, log a message
Promise.resolve().then(function() {
  console.log('Promise perf:', performance.now() - start);
});
```

vs

```js
var start = performance.now();

// Mimic a synchronous render that completes immediately
var render = function() {
  log();
};
var log = function() {
  console.log('sync', performance.now() - start);
};

render();
```

I ran these a few times in the console on this page, and the synchronous version is on the order of 0.015ms, and the async version is on the order of .9ms.

Those other libraries are faster than us because of the other technologies they use to manage the DOM, like incremental updates & run loops. At this time I think that we're unlikely to add those to Mn core.

**Comment by somebody32** [comment]
@jmeas you're right, but async rendering allows you to unblock the main thread, so user will not be forced to see the frozen UI. Also, you can combine it with the bathched DOM inserting, so page will render progressively, without need to wait for all N items in the collection to render

**Comment by jmeas** [comment]
Right, for sure. I was grouping all of those fancy things in with the "other technologies" I mentioned at the bottom of my post.

**Comment by megawac** [comment]
I feel like both those arent fancy but natural benefits of async render
strategies

**Comment by Tvrqvoise** [comment]
Asynchronous render is slower if you measure JS execution alone.  For running a series of synchronous steps, nothing can best synchronous execution.  Asynchronous render only has the potential to be faster if you're optimizing for the network and browser paints, which I feel is a more likely bottleneck.

**Comment by paulfalgout** [comment]
This renderer is external for v4
